pkg/route/validation/validation_test.go

```diff
@@ -843,6 +978,717 @@ func TestValidateRoute(t *testing.T) {
+										Type: routev1.Set,
+										Set: &routev1.RouteSetHTTPHeader{
+											Value: "text/plain,text/html",
+										},
+									},
+								},
+								{
+									Name: "Accept",
+									Action: routev1.RouteHTTPHeaderActionUnion{
+										Type: routev1.Delete,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedErrorMessage: `spec.httpHeaders.actions.request[1].name: Duplicate value: "Accept"`,
+		},
+		{
+			name: "set is required when type is Set",
+			route: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "missing-set-field",
+					Namespace: "foo",
+				},
+				Spec: routev1.RouteSpec{
+					Host: "subdomain.example.test",
+					To:   createRouteSpecTo("serviceName", "Service"),
+					HTTPHeaders: &routev1.RouteHTTPHeaders{
+						Actions: routev1.RouteHTTPHeaderActions{
+							Request: []routev1.RouteHTTPHeader{
+								{
+									Name: "Accept",
+									Action: routev1.RouteHTTPHeaderActionUnion{
+										Type: routev1.Set,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedErrorMessage: `spec.httpHeaders.actions.request[0].action.set: Required value: set is required when type is Set, and forbidden otherwise`,
+		},
+		{
+			name: "set is forbidden when type is not Set",
+			route: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "missing-set-field",
+					Namespace: "foo",
+				},
+				Spec: routev1.RouteSpec{
+					Host: "subdomain.example.test",
+					To:   createRouteSpecTo("serviceName", "Service"),
+					HTTPHeaders: &routev1.RouteHTTPHeaders{
+						Actions: routev1.RouteHTTPHeaderActions{
+							Request: []routev1.RouteHTTPHeader{
+								{
+									Name: "Accept",
+									Action: routev1.RouteHTTPHeaderActionUnion{
+										Type: routev1.Delete,
+										Set: &routev1.RouteSetHTTPHeader{
+											Value: "text/plain,text/html",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedErrorMessage: `spec.httpHeaders.actions.request[0].action.set: Required value: set is required when type is Set, and forbidden otherwise`,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var allErrs field.ErrorList
+			allErrs = append(allErrs, validateHeaders(field.NewPath("spec", "httpHeaders", "actions", "response"), tc.route.Spec.HTTPHeaders.Actions.Response, permittedResponseHeaderValueRE, permittedResponseHeaderValueErrorMessage)...)
+			allErrs = append(allErrs, validateHeaders(field.NewPath("spec", "httpHeaders", "actions", "request"), tc.route.Spec.HTTPHeaders.Actions.Request, permittedRequestHeaderValueRE, permittedRequestHeaderValueErrorMessage)...)
+			var actualErrorMessage string
+			if err := allErrs.ToAggregate(); err != nil {
+				actualErrorMessage = err.Error()
+			}
+			switch {
+			case tc.expectedErrorMessage == "" && actualErrorMessage != "":
+				t.Fatalf("unexpected error: %v", actualErrorMessage)
```

Comment:
I think all of these should be `Errorf`, not `Fatalf`

Reply:
Does it matter?  We're using `t.Run` here, and there are no following statements in the function.

Reply:
But we range over a set of tests.  Doesn't `Fatalf` stop and return, thus ending the test suite, rather than finishing the loop?

Reply:
No, `Fatalf` only terminates the inner function, not the outer loop.  I wrote a short demonstration here: https://go.dev/play/p/m4jbVtHM9BC